Crash in WebKit::RemoteLayerBackingStore::needsDisplay due to owner() being null.

mattwoodrow opened the pull request:
#### 69f43e22401674b905bb6762846ce2f1814146be
<pre>
Crash in WebKit::RemoteLayerBackingStore::needsDisplay due to owner() being null.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276352">https://bugs.webkit.org/show_bug.cgi?id=276352</a>
&lt;<a href="https://rdar.apple.com/131107394">rdar://131107394</a>&gt;

Reviewed by Dan Glastonbury.

When a layer subtree has rendering suppression, we just want add its backing store
to m_reachableBackingStoreInLatestFlush to stop it from being purged.

It is entirely possible for the owner() (the GraphicsLayer) to have gone away, and
we want to continue to display the old content until suppression ends.

The existing code for RemoteLayerBackingStoreCollection::backingStoreWillBeDisplayed
checked for invalidation via needsDisplay(), which then crashed if the owner had gone
away.

This adds a new alternative that doesn&apos;t check for invalidation, with very explicit
rendering suppression naming.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::layerWillBeDisplayedWithRenderingSuppression):
(WebKit::RemoteLayerBackingStore::needsDisplay const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeDisplayedWithRenderingSuppression):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::recursiveMarkWillBeDisplayedWithRenderingSuppresion):
(WebKit::PlatformCALayerRemote::recursiveBuildTransaction):
(WebKit::PlatformCALayerRemote::recursiveMarkWillBeDisplayed): Deleted.

Canonical link: <a href="https://commits.webkit.org/280874@main">https://commits.webkit.org/280874@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d469ad2a3ba22584a4e82e4e936a02e7f31f119

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57911 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37239 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10387 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8356 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8544 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5947 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59941 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/34928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/50054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/27758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/31694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7360 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/7617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63220 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1825 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7687 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1832 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/50065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/54289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/1557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8633 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33068 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34154 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35238 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33899 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -317,7 +328,7 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
     }
 
-    if (m_layer->owner()->platformCALayerDelegatesDisplay(m_layer)) {
+    if (m_layer->owner() && m_layer->owner()->platformCALayerDelegatesDisplay(m_layer)) {
```
Is there any chance that `m_layer->owner()` could be nulled between `m_layer->owner()` and `m_layer->owner()->`? Would it be better to place the result of `m_layer->owner()` in a temporary variable?

mattwoodrow replied:
I don't think so, but I agree with the pattern :)

I might actually remove this change, as the other changes mean that we never end up here without an owner. Probably better to just consider it invalid to run this code without an owner, rather than silently returning a different answer.

djg reviewed (approved):
:shipit:

mattwoodrow committed:
Crash in WebKit::RemoteLayerBackingStore::needsDisplay due to owner() being null.
https://bugs.webkit.org/show_bug.cgi?id=276352
<rdar://131107394>

Reviewed by Dan Glastonbury.

When a layer subtree has rendering suppression, we just want add its backing store
to m_reachableBackingStoreInLatestFlush to stop it from being purged.

It is entirely possible for the owner() (the GraphicsLayer) to have gone away, and
we want to continue to display the old content until suppression ends.

The existing code for RemoteLayerBackingStoreCollection::backingStoreWillBeDisplayed
checked for invalidation via needsDisplay(), which then crashed if the owner had gone
away.

This adds a new alternative that doesn't check for invalidation, with very explicit
rendering suppression naming.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::layerWillBeDisplayedWithRenderingSuppression):
(WebKit::RemoteLayerBackingStore::needsDisplay const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeDisplayedWithRenderingSuppression):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::recursiveMarkWillBeDisplayedWithRenderingSuppresion):
(WebKit::PlatformCALayerRemote::recursiveBuildTransaction):
(WebKit::PlatformCALayerRemote::recursiveMarkWillBeDisplayed): Deleted.

Canonical link: https://commits.webkit.org/280874@main
